### PR TITLE
NDRS-807: Remove deploy/transfer hashes and proposer out of header

### PR DESCRIPTION
**Pull request by darthsiroftardis**
REF: https://casperlabs.atlassian.net/browse/NDRS-807

CHANGELOG:

- Moved three fields: `deploy_hashes`, `transfer_hashes` and `proposer` out of `BlockHeader` into new stuct `BlockBody`
- Refactored `linear_chain_sync` and `linear_chain_fast_sync` to have `Block` instead of `BlockHeader`
- Refactored `era_supervisor` to have `Block` when announcing handling of linear chain blocks.
- Refactored JsonBlock to account for the change in DTO Block structure.

### CODE REVIEW

File: node/src/reactor/joiner.rs
Comment (dwerner) [inline node/src/reactor/joiner.rs:826]: This is not clear: is it a block or a block_header?

File: node/src/reactor/validator.rs
Comment (dwerner) [inline node/src/reactor/validator.rs:909]: Same as above.

**Review by dwerner — APPROVED** [review]
Looks good to me, I had one nitpick, but assuming that's addressed, I think we need to let @zie1ony know that this affects the event stream serialized type.

File: node/src/reactor/joiner.rs
Comment (darthsiroftardis) [reply node/src/reactor/joiner.rs]: It should be block_header

File: node/src/reactor/validator.rs
Comment (darthsiroftardis) [reply node/src/reactor/validator.rs]: `block_header`

File: node/src/components/consensus/era_supervisor.rs
Comment (goral09) [inline node/src/components/consensus/era_supervisor.rs:496]: Why do we need a whole `Block` here if all you ever use are header's fields?

Comment (afck) [reply node/src/components/consensus/era_supervisor.rs:496]: I think it was because of `announce_block_handled` below.

File: node/src/types/block.rs
Comment (afck) [inline node/src/types/block.rs:753]:
```suggestion
    /// Creates a new body from deploy and transfer hashes.
```
(Method doc comments are usually in the third person; also below.)

File: node/src/types/block.rs
Comment (afck) [inline node/src/types/block.rs:898]: Might be easier to just put the three fields directly into `Block`; I imagine we won't be using the `BlockBody` without the header anywhere?
(Just a thought; I don't have a strong opinion on that.)

**Review by afck — APPROVED** [review]

File: node/src/components/consensus/era_supervisor.rs
Comment (goral09) [reply node/src/components/consensus/era_supervisor.rs:496]: I thought it's needed by the `linear_chain` but it's for `linear_chain_sync`. Makes sense.

**Review by xcthulhu — CHANGES_REQUESTED** [review]
Block bodies and block headers need to stored separately now.

Looks like we have to make the appropriate changes to `components::storage::handle_storage_request` to enable this.

**Comment by goral09** [comment]
> Block bodies and block headers need to stored separately now.
> 
> Looks like we have to make the appropriate changes to `components::storage::handle_storage_request` to enable this.

Can we make this in a separate PR?

**Comment by xcthulhu** [comment]
> Can we make this in a separate PR?

It can't slip, so I'd rather not.

**Comment by goral09** [comment]
> > Can we make this in a separate PR?
> 
> It can't slip, so I'd rather not.

Not that it would slip but this is something that I will need/work on as part of my work. I would like to avoid the potential conflicts. 

It that part of work doesn't have to be made in this PR then I would prefer to leave it out.

File: node/src/components/storage.rs
Comment (goral09) [inline node/src/components/storage.rs:359]: Why we can't have more?

File: node/src/components/storage.rs
Comment (goral09) [inline node/src/components/storage.rs:363]: Aren't we overwriting finality signatures with this?

Comment (goral09) [reply node/src/components/storage.rs:363]: Yes, we are. We can't do that.

**Review by goral09 — CHANGES_REQUESTED** [review]
We cannot reuse `block_metadata_db` b/c we will be overwriting finality signatures.

File: node/src/components/storage.rs
Comment (xcthulhu) [reply node/src/components/storage.rs]: Fixed in 8039fb9ccf0e55a813689f9d44c7fb190e1f42b5

File: node/src/components/storage.rs
Comment (xcthulhu) [reply node/src/components/storage.rs]: Fixed in 8039fb9ccf0e55a813689f9d44c7fb190e1f42b5

**Review by afck — APPROVED** [review]
Looks good to me, but method comments should be third person, and I'd wait for another approval from someone more familiar with storage.

**Review by Fraser999 — APPROVED** [review]
Approved assuming the multiple doc comment changes Andreas asked for are implemented.

File: node/src/components/storage.rs
Comment (goral09) [inline node/src/components/storage.rs:376]: We should try to remove the body if we failed to insert the header.

**Comment by darthsiroftardis** [comment]
@xcthulhu @goral09 Can you please re-view to see if the changes are satisfactory

File: node/src/components/storage.rs
Comment (darthsiroftardis) [reply node/src/components/storage.rs:376]: Right but this is in the case of a failure or some error to store the the block

Comment (goral09) [reply node/src/components/storage.rs:376]: Yes, I understand that. I just want to make sure that we're not leaving any leftover, unnecessary state.

Comment (goral09) [reply node/src/components/storage.rs:376]: I checked. It does revert it.

Comment (darthsiroftardis) [reply node/src/components/storage.rs:376]: Aaaah okay, sorry my mistake

**Review by goral09 — APPROVED** [review]

**Review by xcthulhu — APPROVED** [review]

**Pull request merged**
